src/cdk/schematics/ng-update/upgrade-rules/method-call-arguments-rule.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as ts from 'typescript';
+import {MigrationRule} from '../../update-tool/migration-rule';
+
+import {MethodCallUpgradeData} from '../data';
+import {getVersionUpgradeData, RuleUpgradeData} from '../upgrade-data';
+
+/**
+ * Rule that visits every TypeScript method call expression and checks if the
+ * argument count is invalid and needs to be *manually* updated.
+ */
+export class MethodCallArgumentsRule extends MigrationRule<RuleUpgradeData> {
+  /** Change data that upgrades to the specified target version. */
+  data: MethodCallUpgradeData[] = getVersionUpgradeData(this, 'methodCallChecks');
+
+  // Only enable the migration rule if there is upgrade data.
+  ruleEnabled = this.data.length !== 0;
+
+  visitNode(node: ts.Node): void {
+    if (ts.isCallExpression(node) && ts.isPropertyAccessExpression(node.expression)) {
+      this._checkPropertyAccessMethodCall(node);
+    }
+  }
+
+  private _checkPropertyAccessMethodCall(node: ts.CallExpression) {
+    const propertyAccess = node.expression as ts.PropertyAccessExpression;
+
+    if (!ts.isIdentifier(propertyAccess.name)) {
+      return;
+    }
+
+    const hostType = this.typeChecker.getTypeAtLocation(propertyAccess.expression);
+    const hostTypeName = hostType.symbol && hostType.symbol.name;
+    const methodName = propertyAccess.name.text;
+
+    if (!hostTypeName) {
+      return;
+    }
+
+    // TODO(devversion): Revisit the implementation of this upgrade rule. It seems
+    // difficult and ambiguous to maintain the data for this rule.
```

Comment:
My thinking was that it's confusing to track the breaking changes based on the argument count. This can become ambiguous if a method just changed the type of a parameter and the argument count hasn't changed.

Surely in that case it's also possible to just add an entry for the same argument count, but you'd need to do this multiple times if the function has optional parameters. Also the error message is not helpful, nor would it be easy to detect whether the current argument is assignable to the new type.

We did similar stuff for the signature-checks rule and my thinking was that we could re-use that logic in order to make this rule future-proof. Not a priority for now though (that's why I added a TODO)

Reply:
Updated the comment. Let me know if you think that TODO is not needed. I don't feel very strong about it.